sycl/doc/extensions/experimental/sycl_ext_oneapi_annotated_ptr.asciidoc

```diff
@@ -525,6 +525,7 @@ class annotated_ref {
     annotated_ref(const annotated_ref&) = delete;
     operator T() const;
     T operator=(const T &) const;
+    T operator=(const volatile T &) const;
```

Comment:
I can't think of any standard C++ library routine that takes a `const volatile &` in addition to a `const &`  parameter.  @rolandschulz, do you know of any general C++ policy here?

My guess is that anyone using `volatile` needs to know exactly when the variable is read / written.  Therefore, you wouldn't want to pass a `volatile &` to a library function because you wouldn't know how many times the library reads / writes through the reference.

For users who have a `volatile` variable, why not require them to copy the value to a temporary like this:

```
void foo(volatile T& x) {
  T copy = x;
  annotated_ptr<T, ...> aptr;
  *aptr = copy;
}
```

Reply:
I think in a HW compile circumstance isn't it common to have a volatile pointer being passed to a kernel library? 

Considering we passed a volatile pointer to a sycl kernel, the volatile key word tells the compiler do not optimize away any load/store on this pointer since the external HW might change it silently.

Reply:
I think we need a solution because it is reasonable to expect annotated_ptr/ref to work with volatile but STL doesn't care about that.

I think there is one reasonable alternative to the solution in this PR. Instead of adding the overload we change the (only) assignment operator to pass by value: `T operator=(T) const;`. The compiler should always inline therefore there shouldn't be a performance impact of changing that. Also this isn't a breaking change because (outside of volatile) you can pass the same to either operator.

Both solutions should produce the same code.

I think the only minor downside to the current solution is the extra complexity of the overload. I'm not aware of any downside to the other solution.

Reply:
> I think there is one reasonable alternative to the solution in this PR. Instead of adding the overload we change the (only) assignment operator to pass by value: T operator=(T) const

This seems more reasonable to me.  I assume we would to the same thing here:

```
// OP is: +=, -=, *=, /=, %=, <<=, >>=, &=, |=, ^=
T operatorOP(const T &) const;
```

BTW, why are the assignment operators declared `const`?  Shouldn't they all be non-const?

Reply:
>  I assume we would to the same thing here [compound assignment]

Not just that. We should do this consistently for all arguments taking `const T&`.

>  why are the assignment operators declared const? Shouldn't they all be non-const?

`annotated_ref` doesn't own the data (like e.g. span or ptr). The assignment changes the data not what we refer to. Therefore it is correct to assign to a `const annotated_ref` (which is allowed by it not being const) but it isn't correct to assign to assign to `annoated_ref<const T>`.

Reply:
Updated the PR